Refactor Artifact Structure across the package creation tasks

azlam-abdulsalam opened the pull request:
sfpowerscripts feature artifact as the central idea for connecting build and release. Till Milestone 16, the artifacts were only metadata and it need to be checked out from version control, in the case of source package. Delta package did generate the artifacts, but the package format was not fully thought out.
Also to satisfy, skipping of artifacts, if installed not only in unlocked package, also in source package, it is much simpler if we have the artifact carried as part of the build process. This was already done as part of #101, however we have decided to make it uniform across the entire packages.

Every artifact will be named <package_name>_sfpowerscripts_aritfact or spowerscripts_artifact (only in the case of delta package, generated across all the repositories)

Each artifact will contain an aritfact_metadata.json (which has all the information regaring the artifact). as well as a folder called source (which is the metadata that will be deployed). 

Backward compatibility is not supported, The Install Source Package or Deploy Task will only work with the newer versions of associated artifact producer tasks

Fixes #56 
Fixes #101

Commit:
Refactor Package Artifact Structure

Commit:
Refactor Package Artifact Structure

Commit:
Refactor Artifact File names

Commit:
Refactor Artifact File names

Commit:
Add log to metadata

Commit:
Add log to metadata

Commit:
Fix Package Metadata logging

Commit:
Fix Package Metadata logging

Commit:
Fixes for crashes when missing destructiveManifest

Commit:
Fixes for crashes when missing destructiveManifest

Commit:
Refactor and add error handling to destructiveManifests

Commit:
Refactor and add error handling to destructiveManifests

aly76 commented on packages/core/src/sfdxutils/ArtifactGenerator.ts:
```diff
@@ -0,0 +1,86 @@
+
+      let sourcePackage: string = path.join(
+        sfdx_package_artifact,
+        `source `
```
There's a trailing blank space after source

Commit:
Add logging to figure path issues with delta

Commit:
Add logging to figure path issues with delta

Commit:
fix a bug when package directory is provided for soruce artifact

Commit:
fix a bug when package directory is provided for soruce artifact

Commit:
Fix scenario, when package name is not provided for delta package

Commit:
Fix scenario, when package name is not provided for delta package

Commit:
Fix package requirement in Delta Task

Commit:
Fix package requirement in Delta Task

Commit:
Refactor isEmpty Check to sfdxutils

Commit:
Fix Merge conflicts from upstream

Commit:
Fix deepscan bugs

Commit:
Fix deepscan issues

Commit:
Fix bad destsrutctive manifest on Create Unlocked Package

Commit:
Refactor Generate Source with bugfixes

- Introduce async retry for Org Check

- Fix Delta Package Issues

- Remove Project directory flags, Fix issues with CLI

- Add Profile check, futher refactoring to ManifestHelpers

Commit:
Refactor variable name

Commit:
Add support when all packages are  there in an artifact, such as diff

Commit:
Fix Incorrect logic for artifact check

Commit:
Add logs

Commit:
Fix Incorrect dereferencing

Commit:
Fix bugs with Path

Commit:
fix path to source

Commit:
Fix wrong var in install source

Commit:
Fix Manifest when the directory doesn not exist

Commit:
Fix couple of bugs with delta task when package is not provided

Commit:
Remove unwanted console logs

Commit:
Fix Deepscan issues

Commit:
Add package_type to delta in Delta Package Create

Commit:
Update CLI to switch to delta before writing to file

Commit:
Add Error condition to break Delta, if multiple packages are found

Commit:
Fix Output variables

Commit:
Fix Conflicts with develop

Commit:
Fix deepscan issue
